File: src/golang-backend/api/tenant_api.go
package api

import (
	"encoding/json"
	"net/http"
	"time"

	"github.com/gorilla/mux"
)

// Tenant represents a SaaS tenant/customer
type Tenant struct {
	ID                    string                 `json:"id"`
	Name                  string                 `json:"name"`
	Slug                  string                 `json:"slug"`
	Domain                string                 `json:"domain,omitempty"`
	Status                string                 `json:"status"` // active, suspended, trial, cancelled
	SubscriptionTier      string                 `json:"subscriptionTier"` // free, starter, professional, enterprise
	SubscriptionStartDate string                 `json:"subscriptionStartDate"`
	SubscriptionEndDate   string                 `json:"subscriptionEndDate"`
	MaxUsers              int                    `json:"maxUsers"`
	CurrentUsers          int                    `json:"currentUsers"`
	MaxStorage            int                    `json:"maxStorage"` // in GB
	CurrentStorage        int                    `json:"currentStorage"` // in GB
	Features              []string               `json:"features"`
	BillingEmail          string                 `json:"billingEmail"`
	ContactPerson         string                 `json:"contactPerson"`
	Phone                 string                 `json:"phone"`
	Address               string                 `json:"address,omitempty"`
	Logo                  string                 `json:"logo,omitempty"`
	Metadata              map[string]interface{} `json:"metadata"`
	CreatedAt             string                 `json:"createdAt"`
	UpdatedAt             string                 `json:"updatedAt"`
}

// CreateTenantRequest represents request body for creating a tenant
type CreateTenantRequest struct {
	Name                  string                 `json:"name" validate:"required"`
	Slug                  string                 `json:"slug" validate:"required"`
	Domain                string                 `json:"domain,omitempty"`
	SubscriptionTier      string                 `json:"subscriptionTier" validate:"required,oneof=free starter professional enterprise"`
	SubscriptionEndDate   string                 `json:"subscriptionEndDate" validate:"required"`
	BillingEmail          string                 `json:"billingEmail" validate:"required,email"`
	ContactPerson         string                 `json:"contactPerson" validate:"required"`
	Phone                 string                 `json:"phone" validate:"required"`
	Address               string                 `json:"address,omitempty"`
	Metadata              map[string]interface{} `json:"metadata,omitempty"`
}

// UpdateTenantRequest represents request body for updating a tenant
type UpdateTenantRequest struct {
	Name          string                 `json:"name,omitempty"`
	Status        string                 `json:"status,omitempty" validate:"omitempty,oneof=active suspended trial cancelled"`
	Domain        string                 `json:"domain,omitempty"`
	BillingEmail  string                 `json:"billingEmail,omitempty" validate:"omitempty,email"`
	ContactPerson string                 `json:"contactPerson,omitempty"`
	Phone         string                 `json:"phone,omitempty"`
	Address       string                 `json:"address,omitempty"`
	Metadata      map[string]interface{} `json:"metadata,omitempty"`
}

// TenantAnalytics represents tenant analytics data
type TenantAnalytics struct {
	TotalTenants          int            `json:"totalTenants"`
	ActiveTenants         int            `json:"activeTenants"`
	TrialTenants          int            `json:"trialTenants"`
	SuspendedTenants      int            `json:"suspendedTenants"`
	TotalRevenue          float64        `json:"totalRevenue"`
	MRR                   float64        `json:"mrr"` // Monthly Recurring Revenue
	ARR                   float64        `json:"arr"` // Annual Recurring Revenue
	AverageUsersPerTenant float64        `json:"averageUsersPerTenant"`
	TotalStorageUsed      float64        `json:"totalStorageUsed"`
	SubscriptionBreakdown map[string]int `json:"subscriptionBreakdown"`
}

// UsageMetric represents usage tracking data
type UsageMetric struct {
	TenantID    string  `json:"tenantId"`
	Date        string  `json:"date"`
	ActiveUsers int     `json:"activeUsers"`
	StorageUsed float64 `json:"storageUsed"` // in GB
	APICalls    int     `json:"apiCalls"`
	Bandwidth   float64 `json:"bandwidth"` // in GB
}

// Invoice represents billing invoice
type Invoice struct {
	ID            string `json:"id"`
	TenantID      string `json:"tenantId"`
	InvoiceNumber string `json:"invoiceNumber"`
	Amount        float64 `json:"amount"`
	Currency      string `json:"currency"`
	Status        string `json:"status"` // paid, pending, overdue, cancelled
	BillingPeriod struct {
		Start string `json:"start"`
		End   string `json:"end"`
	} `json:"billingPeriod"`
	Items []InvoiceItem `json:"items"`
	IssueDate string `json:"issueDate"`
	DueDate   string `json:"dueDate"`
	PaidDate  string `json:"paidDate,omitempty"`
}

// InvoiceItem represents a line item in an invoice
type InvoiceItem struct {
	Description string  `json:"description"`
	Quantity    int     `json:"quantity"`
	UnitPrice   float64 `json:"unitPrice"`
	Total       float64 `json:"total"`
}

// TenantHandler handles tenant-related HTTP requests
type TenantHandler struct {
	// Add database connection here
	// db *gorm.DB
}

// NewTenantHandler creates a new tenant handler
func NewTenantHandler() *TenantHandler {
	return &TenantHandler{}
}

// RegisterRoutes registers tenant API routes
func (h *TenantHandler) RegisterRoutes(r *mux.Router) {
	// Tenant CRUD
	r.HandleFunc("/api/tenants", h.GetAllTenants).Methods("GET")
	r.HandleFunc("/api/tenants/{id}", h.GetTenantByID).Methods("GET")
	r.HandleFunc("/api/tenants", h.CreateTenant).Methods("POST")
	r.HandleFunc("/api/tenants/{id}", h.UpdateTenant).Methods("PUT")
	r.HandleFunc("/api/tenants/{id}", h.DeleteTenant).Methods("DELETE")
	
	// Tenant status management
	r.HandleFunc("/api/tenants/{id}/status", h.UpdateTenantStatus).Methods("PATCH")
	r.HandleFunc("/api/tenants/{id}/subscription", h.UpgradeTenantSubscription).Methods("PATCH")
	
	// Usage metrics
	r.HandleFunc("/api/tenants/{id}/usage", h.GetUsageMetrics).Methods("GET")
	r.HandleFunc("/api/tenants/{id}/usage", h.RecordUsageMetric).Methods("POST")
	
	// Invoices
	r.HandleFunc("/api/tenants/{id}/invoices", h.GetTenantInvoices).Methods("GET")
	r.HandleFunc("/api/invoices", h.CreateInvoice).Methods("POST")
	r.HandleFunc("/api/invoices/{id}/status", h.UpdateInvoiceStatus).Methods("PATCH")
	
	// Analytics
	r.HandleFunc("/api/tenants/analytics", h.GetTenantAnalytics).Methods("GET")
}

// GetAllTenants returns all tenants
// @Summary Get all tenants
// @Description Retrieve all tenants with optional filtering
// @Tags Tenants
// @Accept json
// @Produce json
// @Param status query string false "Filter by status (active, trial, suspended, cancelled)"
// @Param tier query string false "Filter by subscription tier"
// @Success 200 {array} Tenant
// @Router /api/tenants [get]
func (h *TenantHandler) GetAllTenants(w http.ResponseWriter, r *http.Request) {
	// Implementation here
	// Query database, apply filters, return results
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]interface{}{
		"message": "Get all tenants",
		"data": []Tenant{},
	})
}

// GetTenantByID returns a specific tenant
// @Summary Get tenant by ID
// @Description Retrieve tenant details by ID
// @Tags Tenants
// @Accept json
// @Produce json
// @Param id path string true "Tenant ID"
// @Success 200 {object} Tenant
// @Failure 404 {object} map[string]string
// @Router /api/tenants/{id} [get]
func (h *TenantHandler) GetTenantByID(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	tenantID := vars["id"]
	
	// Implementation: Query database for tenant
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]interface{}{
		"id": tenantID,
		"message": "Get tenant by ID",
	})
}

// CreateTenant creates a new tenant
// @Summary Create new tenant
// @Description Create a new tenant with subscription
// @Tags Tenants
// @Accept json
// @Produce json
// @Param tenant body CreateTenantRequest true "Tenant data"
// @Success 201 {object} Tenant
// @Failure 400 {object} map[string]string
// @Router /api/tenants [post]
func (h *TenantHandler) CreateTenant(w http.ResponseWriter, r *http.Request) {
	var req CreateTenantRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}
	
	// Validate request
	// Create tenant in database
	// Return created tenant
	
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusCreated)
	json.NewEncoder(w).Encode(map[string]interface{}{
		"message": "Tenant created successfully",
	})
}

// UpdateTenant updates tenant information
func (h *TenantHandler) UpdateTenant(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	tenantID := vars["id"]
	
	var req UpdateTenantRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}
	
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]interface{}{
		"id": tenantID,
		"message": "Tenant updated successfully",
	})
}

// DeleteTenant deletes a tenant
func (h *TenantHandler) DeleteTenant(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	tenantID := vars["id"]
	
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]interface{}{
		"id": tenantID,
		"message": "Tenant deleted successfully",
	})
}

// UpdateTenantStatus updates tenant status
func (h *TenantHandler) UpdateTenantStatus(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	tenantID := vars["id"]
	
	var req struct {
		Status string `json:"status"`
	}
	json.NewDecoder(r.Body).Decode(&req)
	
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]interface{}{
		"id": tenantID,
		"status": req.Status,
		"message": "Tenant status updated",
	})
}

// UpgradeTenantSubscription upgrades tenant subscription
func (h *TenantHandler) UpgradeTenantSubscription(w http.ResponseWriter, r *http.Request) {
	// Implementation
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]interface{}{
		"message": "Subscription upgraded",
	})
}

// GetUsageMetrics returns usage metrics for a tenant
func (h *TenantHandler) GetUsageMetrics(w http.ResponseWriter, r *http.Request) {
	// Implementation
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]interface{}{
		"message": "Usage metrics",
		"data": []UsageMetric{},
	})
}

// RecordUsageMetric records usage metric
func (h *TenantHandler) RecordUsageMetric(w http.ResponseWriter, r *http.Request) {
	// Implementation
	w.WriteHeader(http.StatusCreated)
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]interface{}{
		"message": "Usage metric recorded",
	})
}

// GetTenantInvoices returns invoices for a tenant
func (h *TenantHandler) GetTenantInvoices(w http.ResponseWriter, r *http.Request) {
	// Implementation
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]interface{}{
		"message": "Tenant invoices",
		"data": []Invoice{},
	})
}

// CreateInvoice creates a new invoice
func (h *TenantHandler) CreateInvoice(w http.ResponseWriter, r *http.Request) {
	// Implementation
	w.WriteHeader(http.StatusCreated)
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]interface{}{
		"message": "Invoice created",
	})
}

// UpdateInvoiceStatus updates invoice status
func (h *TenantHandler) UpdateInvoiceStatus(w http.ResponseWriter, r *http.Request) {
	// Implementation
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]interface{}{
		"message": "Invoice status updated",
	})
}

// GetTenantAnalytics returns analytics across all tenants
func (h *TenantHandler) GetTenantAnalytics(w http.ResponseWriter, r *http.Request) {
	// Implementation
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(TenantAnalytics{
		TotalTenants: 0,
		ActiveTenants: 0,
		SubscriptionBreakdown: make(map[string]int),
	})
}

File: src/api/tenantApi.ts
/**
 * Tenant API Layer
 * 
 * LocalStorage-based API for tenant management
 * Updated to use new Tenant schema
 * Production: Use /services/tenants-service.ts instead
 */

import type { Tenant, TenantStatus, TenantTier } from '../data/tenants';
import { mockTenants } from '../data/tenants';

const STORAGE_KEY = 'saas_tenants';

// Initialize localStorage with mock data
function initStorage() {
  if (!localStorage.getItem(STORAGE_KEY)) {
    localStorage.setItem(STORAGE_KEY, JSON.stringify(mockTenants));
  }
}

// === TENANT CRUD ===

export async function getAllTenants(): Promise<Tenant[]> {
  initStorage();
  const data = localStorage.getItem(STORAGE_KEY);
  return data ? JSON.parse(data) : [];
}

export async function getTenantById(id: string): Promise<Tenant | null> {
  const tenants = await getAllTenants();
  return tenants.find(t => t._id === id) || null;
}

export async function createTenant(tenant: Partial<Tenant>): Promise<Tenant> {
  const tenants = await getAllTenants();
  const newTenant: Tenant = {
    _id: `tenant-${Date.now()}`,
    code: tenant.code || '',
    name: tenant.name || '',
    tier: tenant.tier || 'FREE',
    status: tenant.status || 'TRIAL',
    data_region: tenant.data_region || 'ap-southeast-1',
    compliance_level: tenant.compliance_level || 'STANDARD',
    parent_tenant_id: tenant.parent_tenant_id || null,
    path: tenant.path || `/${Date.now()}/`,
    billing_type: tenant.billing_type || 'POSTPAID',
    timezone: tenant.timezone || 'UTC',
    profile: tenant.profile || {},
    settings: tenant.settings || {
      max_users: 10,
      max_storage: 10,
      current_users: 0,
      current_storage: 0,
      mfa_enforced: false,
      sso_enabled: false,
      custom_branding: false,
      api_access: false,
      features: [],
    },
    created_at: new Date().toISOString(),
    updated_at: new Date().toISOString(),
    version: 1,
  };
  tenants.push(newTenant);
  localStorage.setItem(STORAGE_KEY, JSON.stringify(tenants));
  return newTenant;
}

export async function updateTenant(id: string, updates: Partial<Tenant>): Promise<Tenant | null> {
  const tenants = await getAllTenants();
  const index = tenants.findIndex(t => t._id === id);
  if (index === -1) return null;

  tenants[index] = {
    ...tenants[index],
    ...updates,
    updated_at: new Date().toISOString(),
  };
  localStorage.setItem(STORAGE_KEY, JSON.stringify(tenants));
  return tenants[index];
}

export async function deleteTenant(id: string): Promise<boolean> {
  const tenants = await getAllTenants();
  const filtered = tenants.filter(t => t._id !== id);
  if (filtered.length === tenants.length) return false;
  localStorage.setItem(STORAGE_KEY, JSON.stringify(filtered));
  return true;
}

export async function updateTenantStatus(id: string, status: TenantStatus): Promise<Tenant | null> {
  return updateTenant(id, { status });
}

export async function upgradeTenantSubscription(
  id: string,
  tier: TenantTier,
  endDate: string
): Promise<Tenant | null> {
  return updateTenant(id, {
    tier,
    subscription_end_date: endDate,
  });
}

// === USAGE METRICS ===

// Legacy type for backward compatibility
export type SubscriptionTier = 'free' | 'starter' | 'professional' | 'enterprise';
export interface UsageMetric {
  tenantId: string;
  date: string;
  activeUsers: number;
  storageUsed: number;
  apiCalls: number;
  bandwidth: number;
}
export interface Invoice {
  id: string;
  tenantId: string;
  invoiceNumber: string;
  amount: number;
  currency: string;
  status: 'paid' | 'pending' | 'overdue' | 'cancelled';
  billingPeriod: { start: string; end: string };
  items: Array<{ description: string; quantity: number; unitPrice: number; total: number }>;
  issueDate: string;
  dueDate: string;
  paidDate?: string;
}

const USAGE_STORAGE_KEY = 'saas_usage_metrics';
const INVOICE_STORAGE_KEY = 'saas_invoices';

export async function getUsageMetrics(tenantId: string, days: number = 30): Promise<UsageMetric[]> {
  const data = localStorage.getItem(USAGE_STORAGE_KEY);
  const allMetrics: UsageMetric[] = data ? JSON.parse(data) : [];
  
  const cutoffDate = new Date();
  cutoffDate.setDate(cutoffDate.getDate() - days);
  
  return allMetrics.filter(
    m => m.tenantId === tenantId && new Date(m.date) >= cutoffDate
  );
}

export async function recordUsageMetric(metric: UsageMetric): Promise<void> {
  const data = localStorage.getItem(USAGE_STORAGE_KEY);
  const metrics: UsageMetric[] = data ? JSON.parse(data) : [];
  metrics.push(metric);
  localStorage.setItem(USAGE_STORAGE_KEY, JSON.stringify(metrics));
}

// === INVOICES ===

export async function getInvoicesByTenant(tenantId: string): Promise<Invoice[]> {
  const data = localStorage.getItem(INVOICE_STORAGE_KEY);
  const allInvoices: Invoice[] = data ? JSON.parse(data) : [];
  return allInvoices.filter(inv => inv.tenantId === tenantId);
}

export async function createInvoice(invoice: Omit<Invoice, 'id'>): Promise<Invoice> {
  const data = localStorage.getItem(INVOICE_STORAGE_KEY);
  const invoices: Invoice[] = data ? JSON.parse(data) : [];
  
  const newInvoice: Invoice = {
    ...invoice,
    id: `inv-${Date.now()}`,
  };
  
  invoices.push(newInvoice);
  localStorage.setItem(INVOICE_STORAGE_KEY, JSON.stringify(invoices));
  return newInvoice;
}

export async function updateInvoiceStatus(
  id: string,
  status: Invoice['status'],
  paidDate?: string
): Promise<Invoice | null> {
  const data = localStorage.getItem(INVOICE_STORAGE_KEY);
  const invoices: Invoice[] = data ? JSON.parse(data) : [];
  
  const index = invoices.findIndex(inv => inv.id === id);
  if (index === -1) return null;

  invoices[index] = {
    ...invoices[index],
    status,
    ...(paidDate && { paidDate }),
  };
  
  localStorage.setItem(INVOICE_STORAGE_KEY, JSON.stringify(invoices));
  return invoices[index];
}

// === ANALYTICS ===

export interface TenantAnalytics {
  totalTenants: number;
  activeTenants: number;
  trialTenants: number;
  suspendedTenants: number;
  totalRevenue: number;
  mrr: number;
  arr: number;
  averageUsersPerTenant: number;
  totalStorageUsed: number;
  subscriptionBreakdown: Record<SubscriptionTier, number>;
}

export async function getTenantAnalytics(): Promise<TenantAnalytics> {
  const tenants = await getAllTenants();
  const invoiceData = localStorage.getItem(INVOICE_STORAGE_KEY);
  const invoices: Invoice[] = invoiceData ? JSON.parse(invoiceData) : [];
  
  const paidInvoices = invoices.filter(inv => inv.status === 'paid');
  const totalRevenue = paidInvoices.reduce((sum, inv) => sum + inv.amount, 0);
  
  return {
    totalTenants: tenants.length,
    activeTenants: tenants.filter(t => t.status === 'ACTIVE').length,
    trialTenants: tenants.filter(t => t.status === 'TRIAL').length,
    suspendedTenants: tenants.filter(t => t.status === 'SUSPENDED').length,
    totalRevenue,
    mrr: totalRevenue / 12,
    arr: totalRevenue,
    averageUsersPerTenant: tenants.reduce((sum, t) => sum + (t.settings?.current_users || 0), 0) / tenants.length || 0,
    totalStorageUsed: tenants.reduce((sum, t) => sum + (t.settings?.current_storage || 0), 0),
    subscriptionBreakdown: {
      free: tenants.filter(t => t.tier === 'FREE').length,
      starter: tenants.filter(t => t.tier === 'PRO').length,
      professional: tenants.filter(t => t.tier === 'ENTERPRISE').length,
      enterprise: tenants.filter(t => t.tier.startsWith('PARTNER_')).length,
    },
  };
}